indispensable-asn1/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/Asn1Integer.kt

```diff
@@ -0,0 +1,365 @@
+ * Hence, it directly interoperates with [Kotlin MP BigNum](https://github.com/ionspin/kotlin-multiplatform-bignum) and the JVM BigInteger.
+ */
+@Serializable(with = Asn1IntegerSerializer::class)
+sealed class Asn1Integer(internal val uint: VarUInt, private val sign: Sign) {
+
+    enum class Sign {
+        POSITIVE,
+        NEGATIVE
+    }
+
+    override fun toString(): String = when (sign) {
+        Sign.POSITIVE -> uint.toString()
+        Sign.NEGATIVE -> "-${uint}"
+    }
+
+    abstract fun twosComplement(): ByteArray
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Asn1Integer) return false
+
+        if (uint != other.uint) return false
+        if (sign != other.sign) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = uint.hashCode()
+        result = 31 * result + sign.hashCode()
+        return result
+    }
+
+    class Positive internal constructor(uint: VarUInt) : Asn1Integer(uint, Sign.POSITIVE) {
+        override fun twosComplement(): ByteArray = uint.bytes.let {
+            if (it.first().countLeadingZeroBits() == 0) listOf(0.toUByte()) + it else it
+        }.toUByteArray().toByteArray()
+    }
+
+    class Negative internal constructor(uint: VarUInt) : Asn1Integer(uint, Sign.NEGATIVE) {
+        override fun twosComplement(): ByteArray {
+            if (uint == VarUInt(1u)) return byteArrayOf(-1)
+
+            return VarUInt(
+                uint.inv().toString().toMutableList().decimalPlus(listOf('1')).joinToString(separator = "")
+            ).bytes.let {
+                val diff = uint.bytes.size - it.size
+                val list = if (diff == 0) it else mutableListOf(0.toUByte()) + it
+                if (list.first().toByte() >= 0) listOf((-1).toUByte()) + list
+                else it
+            }.toUByteArray().toByteArray()
+        }
+    }
+
+    companion object {
+        val ONE = Asn1Integer.Positive(VarUInt(1u))
+        val ZERO = Asn1Integer.Positive(VarUInt(0u))
+
+        fun fromDecimalString(input: String): Asn1Integer {
+            if (input.matches(REGEX_BASE10)) return Positive(VarUInt(input))
+            if (input.first() == '-' && input.substring(1).matches(REGEX_BASE10))
+                return Negative(VarUInt(input.substring(1)))
+            else throw IllegalArgumentException("NaN: $input")
+        }
+
+        fun fromTwosComplement(input: ByteArray): Asn1Integer =
+            if (input.first() < 0) {
+                Negative(
+                    VarUInt(
+                        VarUInt(input).inv().toString().toMutableList().decimalPlus(listOf('1'))
+                            .joinToString(separator = "")
+                    )
+                )
+            } else Positive(VarUInt(input))
+    }
+}
+
+
+@JvmInline
+internal value class VarUInt(private val words: MutableList<UByte> = mutableListOf(0u)) {
```

Comment:
is this mutable purely so that `trim()` can work? (depending on the answer i might have improvements)

Reply:
yes

Reply:
#185